buildpacks/python/python/validation.go

```diff
@@ -13,117 +12,92 @@ import (
+		// NOTE: need uncached build layer so validation runs for rebuilds
 		Build: true,
+		Cache: false,
 	})
 
-	return fvl
+	return fv
 }
 
-func (i FunctionValidationLayer) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
-	i.layerContributor.Logger = i.logger
+func (fv *FunctionValidator) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
+	fv.layerContributor.Logger = fv.logger
 
-	return i.layerContributor.Contribute(layer, func() (libcnb.Layer, error) {
-		i.logger.Body("Validating function")
+	return fv.layerContributor.Contribute(layer, func() (libcnb.Layer, error) {
+		fv.logger.Body("Validating function")
 
 		var pythonPath []string
-		if i.Invoker.PythonPath() != "" {
-			pythonPath = append(pythonPath, i.Invoker.PythonPath())
+		if fv.invoker.PythonPath() != "" {
+			pythonPath = append(pythonPath, fv.invoker.PythonPath())
 		}
 
-		if i.InvokerDep.PythonPath() != "" {
-			pythonPath = append(pythonPath, i.InvokerDep.PythonPath())
+		if fv.invokerDeps.PythonPath() != "" {
+			pythonPath = append(pythonPath, fv.invokerDeps.PythonPath())
 		}
 
 		if env, found := os.LookupEnv("PYTHONPATH"); found {
 			pythonPath = append(pythonPath, env)
 		}
 
-		buffer := bytes.NewBuffer(nil)
-		cmd := exec.Command("python", "-m", "pyfunc", "check", "-s", i.applicationPath)
-		// cmd := exec.Command("env")
+		cmd := exec.Command("python", "-m", "pyfunc", "check", "-s", fv.applicationPath)
 		cmd.Env = append(os.Environ(), fmt.Sprintf("PYTHONPATH=%s", strings.Join(pythonPath, string(os.PathListSeparator))))
-		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", EnvModuleName, i.module), fmt.Sprintf("%s=%s", EnvFunctionName, i.function))
-		cmd.Stderr = buffer
-		cmd.Stdout = buffer
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", EnvModuleName, fv.module), fmt.Sprintf("%s=%s", EnvFunctionName, fv.function))
 
-		if err := cmd.Run(); err != nil {
-			return layer, fmt.Errorf("%v: %v", buffer.String(), err)
+		if output, err := fv.commandRunner.Run(cmd); err != nil {
+			return layer, fmt.Errorf("%v: %v", output, err)
 		}
 
-		i.logger.Debug("Function was successfully parsed")
+		fv.logger.Debug("Function was successfully parsed")
 		return layer, nil
 	})
 }
 
-func (i FunctionValidationLayer) Name() string {
-	return i.layerContributor.Name
-}
-
-func WithValidationLogger(logger bard.Logger) FunctionValidationOpts {
-	return func(vl *FunctionValidationLayer, metadata map[string]string) {
-		vl.logger = logger
-	}
-}
-
-func WithValidationFunctionClass(functionClass string, override bool) FunctionValidationOpts {
-	return func(vl *FunctionValidationLayer, metadata map[string]string) {
-		vl.override = override
-
-		fSplit := strings.Split(functionClass, ".")
-		if override || (vl.module == "" && vl.function == "") {
-			vl.module = fSplit[0]
-			vl.function = fSplit[1]
-			metadata[EnvModuleName] = fSplit[0]
-			metadata[EnvFunctionName] = fSplit[1]
-		}
-	}
-}
-
-func WithValidationFunctionEnvs(envs map[string]any) FunctionValidationOpts {
```

Comment:
In going to write the tests for this, I came across what I believe to be a simplification. Correct me if I'm wrong, but this class doesn't seem to need to worry about "override" functionality. The only function that it can know to check is based on the function module/name given to it. It shouldn't need to worry about whether the value was a default or not, like the `Function` layer contributor does.

For instance, if the function module/name given is a default. i.e. "override" is false, this would _still_ need to validate it, because there is no telling what function could be run at launch

Reply:
I do notice that the previous validator implementation supported overriding the passed in function module/name via the `MODULE_NAME` and `FUNCTION_NAME` env vars from `func.yaml`. This seems inconsistent with `function.go` which does not allow this. Is my understanding correct in that this is an inconsistency that should be corrected, or is there something I'm missing? If the former, which way should we go? (either make the function layer contributor support overriding, or remove that functionality from the validator [which is what's in this PR currently]).

Reply:
The layer should be running whether something is defined (or not) for `BP_FUNCTION` or in `func.yaml`.

`function.go`'s module precedence:
1. `BP_FUNCTION` if set (override is set to true)
1. `func.yaml`'s `MODULE_NAME` and `FUNCTION_NAME` (Overrides the default env)
1. `BP_FUNCTION` default (nothing to override it so here we are)

Is that not the overriding behaviour you're seeing?

Perhaps to simplify the layers, we should process what the function name and module should be once at the start of the build. These names can be passed to each layer that needs it instead of having each layer translate it.

Reply:
I don't recall the history on this fully, so I'll defer this to @andrew-su. 

Overriding via environment variables was intentional for a previous feature ask though. IIRC `function.go` does also override, though?

Reply:
Essentially, `MODULE_NAME` and `FUNCTION_NAME` is an implementation detail of the invoker and isn't a part of our buildpack's API. They may choose to use it, but can't guarantee it'll still be the same or around in the future.

Reply:
@bbtong The reason it comes into play in this PR is that I'm writing tests for it and it's currently very complex. I didn't want to invest the time to write tests for it if it is actually behavior we don't want to enforce.

Reply:
@andrew-su In looking at the code on `main` and also experimenting locally, this is how I understand the current behavior (before this PR):

`function.go` will take in the function class (usually given by `BP_FUNCTION`), separate it at the `.`, and store it in the launch envs as `MODULE_NAME` and `FUNCTION_NAME` (setting it as either a default or an override).

`validation.go` will also accept the same function class (via `BP_FUNCTION`), separate it at the `.`, and store it in its struct. Additionally though, it will allow the user to override the module/name via `MODULE_NAME` and `FUNCTION_NAME` if specified in `func.yaml`.

This additional behavior is what I'm flagging as inconsistent and unnecessary (and makes testing harder/more complex).

A concrete example: if `BP_FUNCTION` is unset and `func.yaml` defines `MODULE_NAME=func` `FUNCTION_NAME=other`, `function.go` will use `func.main` (the default) and `validation.go` will validate `func.other`.

Reply:
@bbtong I do understand that its arguable that this is a potential bug and should be in a separate PR. From a practical standpoint, it was highly related to the new tests I was adding. But if we'd like to separate this discussion out to a separate PR, I can skip testing `validation.go` for now and do that refactor/testing as a separate PR. Thoughts?

Reply:
Spoke with @andrew-su to clarify expected behavior. Will address any bugs (if there actually is a bug) in a later PR. We also talked about just removing the need for the `MODULE_NAME` and `FUNCTION_NAME` env vars via invoker refactors, but we can discuss later.